### PR TITLE
Topic/initial

### PR DESCRIPTION
**Pull request by HuwCampbell**

### CODE REVIEW

File: DESCRIPTION [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+Type: Package
+Title: Property Testing for R
+Version: 0.1
+Date: 2017-08-14
+Authors@R: person("Huw", "Campbell", email = "huw.campbell@gmail.com",
+                  role = c("aut", "cre"))
+Description: Provides a modern property-based
+  testing system for R, in the spirit of QuickCheck.
+License: MIT + file LICENSE
+Imports: purrr
```
Comment (moodmosaic) [inline DESCRIPTION:11]: I'm new to R and [purrr](http://purrr.tidyverse.org/) looks interesting 😃

**Review by moodmosaic — APPROVED** [review]
[It's an overwhelming PR](https://twitter.com/iamdevloper/status/397664295875805184) to read through... but the general idea looks good! 👍 💯 :1st_place_medal: 🚀 

From what I've seen, you're not using the concept of Range type and combinators for controlling the scope of generated numbers and collections.

/cc @jystic

**Comment by HuwCampbell** [comment]
Cheers.

Yeah I didn't include a Range type. I didn't feel that it was quite as useful in this port because as a dynamic language R could accept either a gen or a constant for a scaling parameter; and because it already contains excellent support for sampling distributions, so things like `Range.linear` are a bit redundant.

**Comment by HuwCampbell** [comment]
@hadley This library extends `testthat` with property based testing including integrated shrinking.

Would you care to give some thoughts before we think about uploading to CRAN?

**Comment by hadley** [comment]
Do you have a readme or something I can look at?

**Comment by HuwCampbell** [comment]
Thanks for responding.

Yes absolutely, there's one for this package inside [this PR](https://github.com/hedgehogqa/r-hedgehog/blob/5df64b3616e74606c971ec7e636c8f7373f72120/README.md), as well as a few vignettes.

[Here](https://github.com/hedgehogqa/r-hedgehog/tree/topic/initial) is the view of the R package.

File: R/state.R [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+#'   state machine generator.
+#' @param commands the list of commands which
+#'   we can select choose from. Only commands
+#'   appropriate for the state will actually be
+#'   selected.
+#'
+#' @importFrom purrr partial
+#'
+#' @return a list of actions to run during testing
+gen.actions <- function(initial.state, commands) {
```
Comment (HuwCampbell) [inline R/state.R:220]: This gen could probably use some improvements. Its API should be more like `gen.list` and `gen.c`.

File: R/property.R [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+  test     <- 0
+
+  # R doesn't have good tail call optimisation, hence the while loop.
+  while ( test < tests ) {
+    # Check if we have discarded too many trials
+    if (discards >= discard.limit)
+      return (fail( message = "discard limit reached"))
+
+    # Run the test with error handling
+    tree   <- gen.run( generator, size )
```
Comment (HuwCampbell) [inline R/property.R:77]: One bit of clunkiness here, at the moment discards aren't caught in the generator; only the property.

File: DESCRIPTION (hunk 1 above)
Comment (HuwCampbell) [reply DESCRIPTION:11]: It is pretty cool yeah. Unfortunately though the `partial` function appears to be quite inefficient. I have timed state machine generation with and without it and it appears to be contributing about half the run time.

**Comment by jacobstanley** [comment]
👍 happy for you to go ahead and merge this, it all seems pretty reasonable

I don't mind if you want to work through the issues raised in later PRs

**Comment by HuwCampbell** [comment]
Cool will do. Thanks for the input all.

**Pull request merged**